src/node_snapshotable.cc

```diff
@@ -129,84 +677,107 @@ void SnapshotBuilder::Generate(SnapshotData* out,
+        if (per_process::cli_options->build_snapshot) {
 #if HAVE_INSPECTOR
-        env->InitializeInspector({});
+          // TODO(joyeecheung): move this before RunBootstrapping().
+          env->InitializeInspector({});
 #endif
-        TryCatch bootstrapCatch(isolate);
-        // TODO(joyeecheung): we could use the result for something special,
-        // like setting up initializers that should be invoked at snapshot
-        // dehydration.
-        MaybeLocal<Value> result =
-            LoadEnvironment(env, StartExecutionCallback{});
-        if (bootstrapCatch.HasCaught()) {
-          PrintCaughtException(isolate, context, bootstrapCatch);
+          // TODO(joyeecheung): we could use the result for something special,
+          // like setting up initializers that should be invoked at snapshot
+          // dehydration.
+          MaybeLocal<Value> result =
+              LoadEnvironment(env, StartExecutionCallback{});
+          if (bootstrapCatch.HasCaught()) {
+            // TODO(joyeecheung): fail by exiting with a non-zero exit code.
+            PrintCaughtException(isolate, main_context, bootstrapCatch);
+            abort();
+          }
+          result.ToLocalChecked();
+          // FIXME(joyeecheung): right now running the loop in the snapshot
+          // builder seems to introduces inconsistencies in JS land that need to
+          // be synchronized again after snapshot restoration.
+          int exit_code = SpinEventLoop(env).FromMaybe(1);
+          CHECK_EQ(exit_code, 0);
+          if (bootstrapCatch.HasCaught()) {
+            // TODO(joyeecheung): fail by exiting with a non-zero exit code.
+            PrintCaughtException(isolate, main_context, bootstrapCatch);
+            abort();
+          }
         }
-        result.ToLocalChecked();
-        // FIXME(joyeecheung): right now running the loop in the snapshot
-        // builder seems to introduces inconsistencies in JS land that need to
-        // be synchronized again after snapshot restoration.
-        int exit_code = SpinEventLoop(env).FromMaybe(1);
-        CHECK_EQ(exit_code, 0);
-        if (bootstrapCatch.HasCaught()) {
-          PrintCaughtException(isolate, context, bootstrapCatch);
-          abort();
+
+        if (per_process::enabled_debug_list.enabled(
+                DebugCategory::MKSNAPSHOT)) {
+          env->PrintAllBaseObjects();
+          printf("Environment = %p\n", env);
         }
-      }
 
-      if (per_process::enabled_debug_list.enabled(DebugCategory::MKSNAPSHOT)) {
-        env->PrintAllBaseObjects();
-        printf("Environment = %p\n", env);
+        // Serialize the native states
+        out->env_info = env->Serialize(&creator);
+        // Serialize the context
+        size_t index = creator.AddContext(
+            main_context, {SerializeNodeContextInternalFields, env});
+        CHECK_EQ(index, SnapshotBuilder::kNodeMainContextIndex);
+
+        // Regenerate all the code cache.
```

Comment:
It is already in the build time snapshot, I am thinking about resolving the TODO I put in the C++ code serialization before opening a PR for that (because I now realize that mkcodecache being separated from mksnapshot is unsafe - at least for the time being)